ejb3/src/main/java/org/jboss/as/ejb3/timerservice/persistence/database/DatabaseTimerPersistence.java

```diff
@@ -1099,11 +1099,11 @@ public void run() {
                                 }
                             }
 
-                            synchronized (DatabaseTimerPersistence.this) {
-                                Set<String> timers = knownTimerIds.get(timedObjectId);
-                                for (String timer : existing) {
-                                    TimerImpl timer1 = timerService.getTimer(timer);
-                                    if (timer1 != null && timer1.getState() != TimerState.CREATED) {
+                            Set<String> timers = knownTimerIds.get(timedObjectId);
```

Comment:
All other access to knownTimerIds in this class is guarded by 'synchronized (DatabaseTimerPersistence.this)' (or synchronized (this) in DatabaseTimerPersistence itself.

Except the first lie of loadActiveTimers, which is probably a bug.

Changing the concurrency handling for just one call seems risky.

The sets in the map also aren't thread safe.

Reply:
D'oh, thanks for catching that Brian.  Added the knownTImerIds back within a sync there.

Reply:
Any further thoughts on this @bstansberry @tadamski ?